Mathlib/Data/Polynomial/Derivation.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2023 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard, Richard Hill
+-/
+import Mathlib.Data.Polynomial.Derivative
+import Mathlib.RingTheory.Derivation.Basic
+import Mathlib.Data.Polynomial.AlgebraMap
+
+/-!
+# Derivations of univariate polynomials
+
+In this file we prove that an `R`-derivation of `Polynomial R` is determined by its value on `X`.
+We also provide a constructor `Polynomial.mkDerivation` that
+builds a derivation from its value on `X`, and a linear equivalence
+`Polynomial.mkDerivationEquiv` between `A` and `Derivation (Polynomial R) A`.
+-/
+
+namespace Polynomial
+
+open scoped BigOperators
+
+noncomputable section
+
+variable {R A : Type _} [CommSemiring R]
+
+/-- ## Derivative as a derivation -/
+
+def derivative' : Derivation R R[X] R[X] where
+  toFun := derivative
+  map_add' _ _ := derivative_add
+  map_smul' := derivative_smul
+  map_one_eq_zero' := derivative_one
+  leibniz' f g := by simp [mul_comm, add_comm, derivative_mul]
+
+variable [AddCommMonoid A] [Module R A] [Module (Polynomial R) A]
```

Comment:
Indexing a typeclass on a non-carrier type will eventually cause pain.

For example, we might one day have another ring `S` with `R ≃+* S` and wish to use some result which includes the term `Module (Polynomial R) A` in a proof that discusses `Module (Polynomial S) A`.

I wonder could we avoid doing this by introducing a `PolynomialAlgebra` typeclass along the lines of:
```lean
class PolynomialAlgebra (R P : Type _) [CommSemiring R] [Semiring P] extends Algebra R P :=
  (X : P)
  -- TODO add axioms that are equivalent to the map that sends `Polynomial.X` to `PolynomialAlgebra.X` extending to a full algebra equivalence: `R[X] ≃ₐ[R] P`
```

Alternatively we could introduce a typeclass for this specific situation since we know a module over `R[X]` is "the same as" a module over `R` together with an `R`-linear endomorphism (scalar action of `X`):
```lean
class PolynomialModule (R M : Type _) [Semiring R] [AddCommMonoid M] extends Module R M :=
  (X : M →ₗ[R] M)
```

Reply:
Note that @kbuzzard is just copying what we already did for `MvPolynomial` in this PR, not that that invalidates your comment.

Reply:
Good point, I think we can live with this for now but I predict one day we'll want something like `PolynomialAlgebra` or `PolynomialModule`.